src/main/java/games/root/actions/extended/BattleAction.java

```diff
@@ -0,0 +1,586 @@
+                            gs.logEvent(Event.GameEvent.GAME_EVENT, gs.getPlayerFaction(targetPlayerID).toString() + " is now Hostile");
+                        }
+                    }
+                } else if (action instanceof TakeHit td) {
+                    defenderDamage--;
+                    gs.addGameScorePlayer(playerID,1);
+                    if (td.tokenType == RootParameters.TokenType.Sympathy){
+                        playerDestroyedSympathyID = playerID;
+                        lastStage = Stage.removePiecesDefender;
+                        stage = Stage.Outrage;
+                        break;
+                    }
+                } else if (action instanceof DamageItem) {
+                    defenderDamage--;
+                } else if (action instanceof Pass) {
+                    done = true;
+                    lastStage = Stage.removePiecesDefender;
+                }
+                if (defenderDamage==0) {
+                    done = true;
+                }
+            case Outrage:
+                if (action instanceof Pass){
+                    stage = Stage.OutrageWoodland;
+                    ArrayList<boolean[]> handVisibility = new ArrayList<>();
+                    for (int i = 0; i <  gs.getPlayerHand(playerDestroyedSympathyID).getSize(); i++) {
+                        boolean[] cardVisibility = new boolean[gs.getNPlayers()];
+                        cardVisibility[playerDestroyedSympathyID] = true;
+                        cardVisibility[gs.getFactionPlayerID(RootParameters.Factions.WoodlandAlliance)] = true;
+                        handVisibility.add(cardVisibility);
+                    }
+                    gs.getPlayerHand(playerDestroyedSympathyID).setVisibility(handVisibility);
+                }else{
+                    stage = lastStage;
+                }
+                break;
+            case OutrageWoodland:
+                stage = lastStage;
+                break;
+
+
+        }
+    }
+
+    @Override
+    public boolean executionComplete(AbstractGameState state) {
+        return done;
+    }
+
+    @Override
+    public BattleAction copy() {
+        BattleAction copy = new BattleAction(playerID);
+        copy.stage = stage;
+        copy.playerDestroyedSympathyID = playerDestroyedSympathyID;
+        copy.targetPlayerID = targetPlayerID;
+        copy.done = done;
+        copy.attackerDamage = attackerDamage;
+        copy.defenderDamage = defenderDamage;
+        copy.locationID = locationID;
+        copy.lastStage = lastStage;
+        return copy;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        }
+        if (obj instanceof BattleAction ba) {
+            return playerID == ba.playerID;
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash("Battle action", playerID);
+    }
```

Comment:
These equals/hashcode implementations are missing a lot of the fields, including several mutable ones. This seems to be a general issue across the contents of the extended actions package. (The core actions seem to have less of an issue).

Reply:
Fixed!